File: shared/src/main/java/ru/privatenull/pnauth/config/AuthSettings.java
package ru.privatenull.pnauth.config;

import java.time.Duration;
import java.util.regex.Pattern;
import ru.privatenull.pnauth.security.HashAlgorithm;

public record AuthSettings(
        int minPasswordLength,
        int maxPasswordLength,
        int maxLoginAttempts,
        Duration lockoutDuration,
        String usernamePattern,
        HashSettings hashing
) {
    public AuthSettings(
            int minPasswordLength,
            int maxPasswordLength,
            int maxLoginAttempts,
            Duration lockoutDuration,
            int hashIterations
    ) {
        this(minPasswordLength, maxPasswordLength, maxLoginAttempts, lockoutDuration,
                "^[A-Za-z0-9_]{3,16}$", new HashSettings(HashAlgorithm.PBKDF2, hashIterations, 12, 2, 65_536, 1));
    }

    public AuthSettings(
            int minPasswordLength,
            int maxPasswordLength,
            int maxLoginAttempts,
            Duration lockoutDuration,
            int hashIterations,
            String usernamePattern,
            ru.privatenull.pnauth.security.HashAlgorithm hashAlgorithm,
            int bcryptCost,
            int argon2Iterations,
            int argon2Memory,
            int argon2Parallelism
    ) {
        this(minPasswordLength, maxPasswordLength, maxLoginAttempts, lockoutDuration,
                usernamePattern, new HashSettings(hashAlgorithm, hashIterations, bcryptCost, argon2Iterations,
                        argon2Memory, argon2Parallelism));
    }

    public AuthSettings {
        if (minPasswordLength < 1) {
            throw new IllegalArgumentException("minPasswordLength must be positive");
        }
        if (maxPasswordLength < minPasswordLength) {
            throw new IllegalArgumentException("maxPasswordLength must not be smaller than minPasswordLength");
        }
        if (maxLoginAttempts < 1) {
            throw new IllegalArgumentException("maxLoginAttempts must be positive");
        }
        if (lockoutDuration.isNegative() || lockoutDuration.isZero()) {
            throw new IllegalArgumentException("lockoutDuration must be positive");
        }
        if (usernamePattern == null || usernamePattern.isBlank()) {
            throw new IllegalArgumentException("usernamePattern must not be blank");
        }
        if (hashing == null) {
            throw new IllegalArgumentException("hashing must not be null");
        }
        try {
            Pattern.compile(usernamePattern);
        } catch (RuntimeException exception) {
            throw new IllegalArgumentException("usernamePattern is invalid", exception);
        }
    }

    public static AuthSettings defaults() {
        return new AuthSettings(8, 64, 5, Duration.ofSeconds(60),
                "^[A-Za-z0-9_]{3,16}$", HashSettings.defaults());
    }

    public int hashIterations() {
        return hashing.pbkdf2Iterations();
    }

    public ru.privatenull.pnauth.security.HashAlgorithm hashAlgorithm() {
        return hashing.algorithm();
    }

    public int bcryptCost() {
        return hashing.bcryptCost();
    }

    public int argon2Iterations() {
        return hashing.argon2Iterations();
    }

    public int argon2Memory() {
        return hashing.argon2Memory();
    }

    public int argon2Parallelism() {
        return hashing.argon2Parallelism();
    }

    public boolean isPasswordValid(String password) {
        return password != null
                && password.length() >= minPasswordLength
                && password.length() <= maxPasswordLength;
    }

    public boolean isUsernameValid(String username) {
        return username != null && Pattern.compile(usernamePattern).matcher(username).matches();
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/config/AuthSettings.kt
package ru.privatenull.pnauth.config

import ru.privatenull.pnauth.security.HashAlgorithm
import java.time.Duration
import java.util.regex.Pattern

@JvmRecord
data class AuthSettings @JvmOverloads constructor(
    val minPasswordLength: Int,
    val maxPasswordLength: Int,
    val maxLoginAttempts: Int,
    val lockoutDuration: Duration,
    val usernamePattern: String,
    val hashing: HashSettings
) {
    constructor(
        minPasswordLength: Int,
        maxPasswordLength: Int,
        maxLoginAttempts: Int,
        lockoutDuration: Duration,
        hashIterations: Int
    ) : this(
        minPasswordLength, maxPasswordLength, maxLoginAttempts, lockoutDuration,
        "^[A-Za-z0-9_]{3,16}$", HashSettings(HashAlgorithm.PBKDF2, hashIterations, 12, 2, 65_536, 1)
    )

    constructor(
        minPasswordLength: Int,
        maxPasswordLength: Int,
        maxLoginAttempts: Int,
        lockoutDuration: Duration,
        hashIterations: Int,
        usernamePattern: String,
        hashAlgorithm: HashAlgorithm,
        bcryptCost: Int,
        argon2Iterations: Int,
        argon2Memory: Int,
        argon2Parallelism: Int
    ) : this(
        minPasswordLength, maxPasswordLength, maxLoginAttempts, lockoutDuration,
        usernamePattern, HashSettings(hashAlgorithm, hashIterations, bcryptCost, argon2Iterations, argon2Memory, argon2Parallelism)
    )

    init {
        require(minPasswordLength >= 1) { "minPasswordLength must be positive" }
        require(maxPasswordLength >= minPasswordLength) { "maxPasswordLength must not be smaller than minPasswordLength" }
        require(maxLoginAttempts >= 1) { "maxLoginAttempts must be positive" }
        require(!lockoutDuration.isNegative && !lockoutDuration.isZero) { "lockoutDuration must be positive" }
        require(usernamePattern.isNotBlank()) { "usernamePattern must not be blank" }
        requireNotNull(hashing) { "hashing must not be null" }
        try {
            Pattern.compile(usernamePattern)
        } catch (exception: RuntimeException) {
            throw IllegalArgumentException("usernamePattern is invalid", exception)
        }
    }

    fun hashIterations(): Int = hashing.pbkdf2Iterations
    fun hashAlgorithm(): HashAlgorithm = hashing.algorithm
    fun bcryptCost(): Int = hashing.bcryptCost
    fun argon2Iterations(): Int = hashing.argon2Iterations
    fun argon2Memory(): Int = hashing.argon2Memory
    fun argon2Parallelism(): Int = hashing.argon2Parallelism

    fun isPasswordValid(password: String?): Boolean {
        return password != null && password.length in minPasswordLength..maxPasswordLength
    }

    fun isUsernameValid(username: String?): Boolean {
        return username != null && Pattern.compile(usernamePattern).matcher(username).matches()
    }

    companion object {
        @JvmStatic
        fun defaults(): AuthSettings {
            return AuthSettings(
                8, 64, 5, Duration.ofSeconds(60),
                "^[A-Za-z0-9_]{3,16}$", HashSettings.defaults()
            )
        }
    }
}
